LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/ModuleConnection/ModuleConnectionHost.cs

```diff
@@ -0,0 +1,207 @@
+                    return new MethodResponse((int)HttpStatusCode.BadRequest);
+                }
+
+
+                Logger.Log(c2d.DevEUI, $"received cloud to device message from direct method: {methodRequest.DataAsJson}", LogLevel.Debug);
+
+                using var cts = methodRequest.ResponseTimeout.HasValue ? new CancellationTokenSource(methodRequest.ResponseTimeout.Value) : null;
+
+                if (await this.classCMessageSender.SendAsync(c2d, cts?.Token ?? CancellationToken.None))
+                {
+                    return new MethodResponse((int)HttpStatusCode.OK);
+                }
+            }
+
+            return new MethodResponse((int)HttpStatusCode.BadRequest);
+        }
+
+        private Task<MethodResponse> ClearCache()
+        {
+            this.loRaDeviceRegistry.ResetDeviceCache();
+
+            return Task.FromResult(new MethodResponse((int)HttpStatusCode.OK));
+        }
+
+        /// <summary>
+        /// Method to update the desired properties.
+        /// We only want to update the auth code if the facadeUri was performed.
+        /// If the update is not acceptable we don't apply it.
+        /// </summary>
+        internal Task OnDesiredPropertiesUpdate(TwinCollection desiredProperties, object userContext)
+        {
+            try
+            {
+                _ = TryUpdateConfigurationWithDesiredProperties(desiredProperties);
+            } catch (ArgumentOutOfRangeException ex)
+            {
+                Logger.Log($"A desired properties update was detected but the parameters are out of range with exception :  {ex}", LogLevel.Warning);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private bool TryUpdateConfigurationWithDesiredProperties(TwinCollection desiredProperties)
+        {
+            if (desiredProperties.Contains(Constants.FacadeServerUrlKey) && !string.IsNullOrEmpty((string)desiredProperties[Constants.FacadeServerUrlKey]))
+            {
+                if (Uri.TryCreate((string)desiredProperties[Constants.FacadeServerUrlKey], UriKind.Absolute, out var uri) && (uri.Scheme == Uri.UriSchemeHttp || uri.Scheme == Uri.UriSchemeHttps))
+                {
+                    this.networkServerConfiguration.FacadeServerUrl = uri;
+                    if (desiredProperties.Contains(Constants.FacadeServerAuthCodeKey))
+                    {
+                        this.networkServerConfiguration.FacadeAuthCode = (string)desiredProperties[Constants.FacadeServerAuthCodeKey];
+                    }
+
+                    Logger.Log("Desired property changed", LogLevel.Debug);
+                    return true;
+                }
+                else
+                {
+                    Logger.Log("The Facade server Url present in device desired properties was malformed.", LogLevel.Error);
+                    throw new ArgumentOutOfRangeException(nameof(desiredProperties));
```

Comment:
`ArgumentOutOfRangeException` is reserved for method arguments whereas here it's being used for a configuration error at run-time.

Reply:
Changed to ConfigurationErrorsException